pmd-core/src/main/java/net/sourceforge/pmd/processor/AbstractPMDProcessor.java

```diff
@@ -103,9 +110,10 @@ public void processFiles(RuleSetFactory ruleSetFactory, List<DataSource> files,
         SourceCodeProcessor processor = new SourceCodeProcessor(configuration);
 
         for (DataSource dataSource : files) {
-            String niceFileName = filenameFrom(dataSource);
+            // this is the real, canonical and absolute filename (not shortened)
+            String realFileName = dataSource.getNiceFileName(false, null);
```

Comment:
should we deprecate this method from the datasource too and remove all such logic from it?

Reply:
@jsotuyod I'm not sure what that would achieve... It looks like we could reimplement everything in terms of DataSource to avoid depending on File like that. In that case, a DataSource needs a way to provide a readable name (eg a URI or something) for reporting.

Reply:
the thing is, not all datasources map to a file. Actually, of all `DataSource` implementation, only one does: `FileDataSource`.

The others take the data from other sources:
* `ReaderDataSource`: from a `Reader`, which is currently used only with data loaded from a database through JDBC
* `StringDataSource`: from a `String`, used mainly by designers and tests, can be used with data from a db through JDBC too
* `ZipDataSource`: from a `ZipEntry` within a zip file

So, when we ask the `DataSource` for the file / filename, it's all lies. The source should have a name, but it's not necessarily a file name, and it should not know how to trim that path to make it short / relative to something else.

Reply:
So possibly the method is badly named (because it's not a *file* name). Yet if a DataSource cannot trim its name, then I don't know what can. As you say a name is not necessarily a file path, and the only thing that knows what the name *means* is the DataSource implementation. So I don't know how relativizing the name *cannot* be left to the implementation.

> Actually, of all DataSource implementation, only one does: FileDataSource.

For a ZipDataSource we can view a zip entry as a file in the filesystem of the zip, but i get the point.

In any case if we want to support autofixes we need to have some "data sources" that are both readable and writable, and possibly also renamable, movable, etc. If we want to support multifile analysis we also need a strong concept of what is a "file".

For the record, IDEs all use an abstraction that allows seeing archive entries, http resources, etc, as files. For example, Intellij OpenAPI's [virtual file system](https://www.jetbrains.org/intellij/sdk/docs/basics/virtual_file_system.html), or eclipse's [IResource](https://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fcore%2Fresources%2FIResource.html) and [IPath](https://help.eclipse.org/mars/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fcore%2Fruntime%2FIPath.html).
Maybe we need something of that kind too.

Reply:
We currently still use it for sorting the files before processing. This means, after this PR, we'll have the logic to determine the short names duplicated. I think, I'll extract it in a helper class used by both....

The DataSource seems like a good name for an abstraction of what PMD is analyzing. However, the rule context e.g. needs a file/filename. As does the rule violation. While DataSource is abstract to work with more than files, anything else doesn't seem to be. That's something we should aim to fix for PMD 7, but not in PMD 6 (unless there is a bug to fix).

Reply:
I've extracted it now into a `ShortFilenameUtil` used by both AbstractRenderer and FileDataSource.

Reply:
I've created #2004 to track this, I'm merging the PR as is for 6.x